apps/performance-tests/src/util/TestReporter.ts

```diff
@@ -0,0 +1,123 @@
+
+  constructor(runner: Mocha.Runner, options: Mocha.MochaOptions) {
+    super(runner, options);
+    this._outputPath = options.reporterOptions?.BENCHMARK_OUTPUT_PATH;
+
+    runner.on(EVENT_SUITE_BEGIN, (suite) => console.log(`\n${suite.title}`));
+    runner.on(EVENT_TEST_BEGIN, (test) => this.onTestStart(test));
+    runner.on(EVENT_TEST_END, (test) => this.onTestEnd(test));
+    runner.on(EVENT_RUN_END, () => {
+      this.printResults();
+      if (this._outputPath && this.failures.length === 0) {
+        this.saveResults();
+      }
+    });
+  }
+
+  /** Run before each test starts. */
+  private onTestStart(test: Mocha.Test) {
+    this._blockHandler.start();
+    process.stdout.write(`${test.title}...`);
+  }
+
+  /** Run after each test passes or fails. */
+  private onTestEnd(test: Mocha.Test) {
+    this._blockHandler.stop();
+
+    const duration = test.duration!;
+    Base.cursor.CR();
+    console.log(`${test.isPassed() ? Base.symbols.ok : Base.symbols.err} ${test.title} (${duration} ms)`);
+
+    const blockingSummary = this._blockHandler.getSummary();
+    this._testInfo.push({
+      test,
+      blockingSummary,
+    });
+  }
+
+  private printResults() {
+    const errors = new Map<string, any>();
+    const results = this._testInfo.map(({ test, blockingSummary }) => {
+      const testName = test.fullTitle();
+      if (test.err) {
+        errors.set(testName, test.err);
+      }
+
+      const blockingInfo = Object.entries(blockingSummary)
+        .filter(([_, val]) => val !== undefined)
+        .map(([key, val]) => `${key}: ${floatToString(val)}`)
+        .join(", ");
+
+      /* eslint-disable @typescript-eslint/naming-convention */
+      return {
+        Status: test.isPassed() ? "PASS" : "FAIL",
+        Test: testName,
+        Duration: `${test.duration!} ms`,
+        Blocks: blockingInfo,
+      };
+      /* eslint-enable @typescript-eslint/naming-convention */
+    });
+
+    console.log();
+    console.log(tableFormatter(results));
+
+    for (const [name, error] of errors) {
+      console.error();
+      console.error(`${name}:`);
+      console.error(error);
+    }
+  }
+
+  /** Saves performance results in a format that is compatible with Github benchmark action. */
+  private saveResults() {
+    const data = this._testInfo.map(({ test, blockingSummary }) => ({
+      name: test.title,
+      unit: "ms",
+      value: test.duration!,
+      extra: Object.entries(blockingSummary)
+        .map(([key, val]) => `${key}: ${val ?? "N/A"}`)
+        .join("\n"),
+    }));
+
+    const outputPath = this._outputPath!;
+    fs.writeFileSync(outputPath, JSON.stringify(data, undefined, 2));
+    console.log(`Test results saved at ${outputPath}`);
+  }
+}
+
+function floatToString(x: number, maxPrecision: number = 2): string {
+  const result = x.toString();
+  const dotPos = result.indexOf(".");
+  return dotPos >= 0 ? result.substring(0, dotPos + maxPrecision + 1) : result;
+}
```

Comment:
why not `.toFixed(2)`?

Reply:
It adds 2 decimals to all numbers, even integers, which I don't really like. Although, I can use `toFixed`  or `toPrecision` for the times and make a special case for `count`.